### DISCUSSION TITLE
Add arrays of arrays definition more simple

### DISCUSSION THREAD

**Discussion by hupo376787** [dotnet/csharplang#4167] [Language Ideas]
## Background and Motivation

Many other languages can use 
`double[][] x = new double[5][3];`
But donet can only do like this:

```
double[][] x = new double[5][];
for (int i = 0; i<= 4; i++)
      x[i] = new int[3];
```

## Proposed API

Add `double[][] x = new double[5][3];`

[a comment by an automated account is not shown]

**Comment by jkotas** [comment]
This is C# limitation. .NET runtime does supports allocating jagged arrays in one shot.

**Reply by hupo376787** [reply]
But C# is in evolution, can we change😊

**Reply by svick** [reply]
I'm curious, how? Which API or IL opcode can do that?

**Reply by hupo376787** [reply]
Java
https://stackoverflow.com/questions/12231453/syntax-for-creating-a-two-dimensional-array

**Reply by jkotas** [reply]
> .NET runtime does supports allocating jagged arrays in one shot.
> Which API or IL opcode can do that?

Arrays have intrinsic constructors provided by runtime magic.

`int[,,,]` has:
 - `.ctor(int, int, int, int)`. This is the constructor that C# calls.
 - `.ctor(int, int, int, int, int, int, int, int)`. Second constructor that provides lower bounds.

`int[][][][]` has constructors that map to this proposal:
- `.ctor(int)`
- `.ctor(int, int)`
- `.ctor(int, int. int)`
- `.ctor(int, int, int, int)`

**Reply by CyrusNajmabadi** [reply]
> But donet can only do like this:

I feel like this is trivially solvable with:

```c#
var array = CreateArray(5, 3);
```

And then just have a helper function that does this.

**Comment by svick** [comment]
If you're in this situation, why wouldn't you use a 2D array? I.e.:

```c#
double[,] x = new double[5, 3];
```

**Comment by spydacarnage** [comment]
Because they're inefficient in comparison to a 1D array of 1D arrays.

**Reply by HaloFour** [reply]
My understanding is that this is no longer true with .NET Core (and was never true with Mono).  Multidimensional arrays _should_ be faster but .NET Framework had a fairly inefficient method of bounds checking.

**Comment by spydacarnage** [comment]
I've seen some recent benchmarks run, and I'm fairly sure they were running
Core 3.1 - the SD arrays were a good 50% quicker.

**Reply by HaloFour** [reply]
Looks like there are still JIT workitems out there to improve their performance:  https://github.com/dotnet/runtime/issues/35293

I'd be curious to see different benchmarks against .NET 5 (or Mono, since I've read a few reports that this performance difference doesn't exist on that runtime).  Either way it seems like an implementation detail that can be fixed by the runtime and not an intrinsic behavior of multidimensional arrays.  Bounds checking aside you'd think that they'd have better performance due to memory locality, since the individual array elements in a jagged array could be anywhere.

My only (minor) concern with syntax like `var foo = new T[x][y];` is that it hides the fact that there are _x_ separate allocations going on.

**Reply by jkotas** [reply]
Some of the overhead of multidimensional arrays is by design and cannot be easily fixed. For example, all multidimensional arrays have lower-bounds. Lower-bounds are very rarely used, but they have to accounted for by the codegen. The existence of lower-bounds also restricts how much optimizations the JIT can do e.g. for bounds-elimination.

**Comment by spydacarnage** [comment]
I've just had it confirmed - it was on 3.1, but he's going to run it again
against NET5 when he gets the chance.
